rest/blip_sync_messages.go

```diff
@@ -0,0 +1,229 @@
+package rest
+
+import (
+	"fmt"
+	"math"
+	"strings"
+	"bytes"
+
+	"github.com/couchbase/go-blip"
+	"github.com/couchbase/sync_gateway/base"
+	"github.com/couchbase/sync_gateway/channels"
+	"github.com/couchbase/sync_gateway/db"
+)
+
+// Function signature for something that generates a sequence id
+type SequenceIDGenerator func() db.SequenceID
+
+// Helper for handling BLIP subChanges requests.  Supports Stringer() interface to log aspects of the request.
+type subChanges struct {
+	rq                    *blip.Message       // The underlying BLIP message
+	logger                base.SGLogger       // A logger object which might encompass more state (eg, blipContext id)
+	sinceZeroValueCreator SequenceIDGenerator // A sequence generator for creating zero'd since values
+}
+
+// Create a new subChanges helper
+func newSubChanges(rq *blip.Message, logger base.SGLogger, sinceZeroValueCreator SequenceIDGenerator) *subChanges {
+	return &subChanges{
+		rq:                    rq,
+		logger:                logger,
+		sinceZeroValueCreator: sinceZeroValueCreator,
+	}
+}
+
+func (s *subChanges) since() db.SequenceID {
+
+	// Depending on the db sequence type, use correct zero sequence for since value
+	sinceSequenceId := s.sinceZeroValueCreator()
+
+	if sinceStr, found := s.rq.Properties["since"]; found {
+		var err error
+		if sinceSequenceId, err = db.ParseSequenceIDFromJSON([]byte(sinceStr)); err != nil {
+			s.logger.LogTo("Sync", "%s: Invalid sequence ID in 'since': %s", s.rq, sinceStr)
+			sinceSequenceId = db.SequenceID{}
```

Comment:
If this is trying to reset to zero, it would need to be using sinceZeroValueCreator().  In general it's a bit hard to follow the zero vs. value handling in subchanges.since() - might want to refactor to invoke sinceZeroValueCreator more intentionally instead of at the top of the function (e.g. when !found, or on error).

Reply:
Fixed in https://github.com/couchbase/sync_gateway/issues/3249